### ISSUE TITLE
[Tracking issue] http delegated routing used in production

### ISSUE THREAD

**Issue by BigLep** [ipfs/kubo#9150]
## Done Criteria
1. HTTP delegated routing is enabled on the ipfs.io gateway, hitting cid.contact in parallel to querying the DHT to determine providers for a CID.  
2. ipfs.io gateway operators are able to assess delegated routing calls health with metrics.

## Notes
This corresponds with "stage 1" in https://www.notion.so/pl-strflt/Indexer-IPFS-Reframe-Q3-Plan-77d0f3fa193b438784d2d85096e315e7

It doesn't matter if STI gives empty results or if the results are unreachable because they're to Filecoin Storage Providers that don't allow data transfer with Bitswap.  Enabling retrieval over Bitswap for Filecoin SP's is a separate effort.

Reframe shipped in Kubo 0.14 as part of https://github.com/ipfs/kubo/issues/8775.  That said, it isn't used in production that we know of.  The purpose of this tracking issue is to track followups that we have already identified as needing to be done or that come up as a result of attempting production deployment.

- [x] https://github.com/ipfs/kubo/issues/9188
- [x] Expose Kubo metrics for reframe calls so a Kubo client has visibility into reframe activity (success/failure, latency). 
- [x] Convert to using the updated HTTP Delegated Routing API: https://github.com/ipfs/go-delegated-routing/issues/63
- [x] Enable on ipfs.io gateway (production usecase): 
   - Old: https://github.com/protocol/bifrost-infra/issues/2120
   - New: https://github.com/protocol/bifrost-infra/issues/2183

----
Expanded scope for items observed while in production:
- [ ] https://github.com/ipfs/go-delegated-routing/issues/55

Other followup issues that should be done, but aren't directly tied to the done criteria.
- [x] https://github.com/ipfs/go-delegated-routing/issues/27
- [x] https://github.com/ipfs/go-delegated-routing/issues/30
- [x] https://github.com/ipfs/ipfs-blog/issues/446
- [ ] https://github.com/protocol/bifrost-infra/issues/2142

**Comment by BigLep** [comment]
2022-08-19 conversation:

The code will live in Kubo.
Docs for this "go package" will live in kubo/docs.  
We can have a doc like "design decisions for routing routers".
Antonio: you could create kubo/docs/exploration-reposts dir, and put it there.
(ipld folks did that in past: https://github.com/ipld/specs/tree/master/design/history/exploration-reports, and then moved content to the proper website doics/specs )

Polish off the design

New routers (3)
- Sequential router
- Parallel router
- Router to filter by method
- May include refractoring
- Includes unit tests
- https://github.com/libp2p/go-libp2p-routing-helpers

Make this work in Kubo
* Config parsing
* Using FX

Error handling (missing methods, missing config)

Config documentation

Sharness smoke test like we did with Reframe

Operations story
* 


Relationship with go-libp2p-routing-helpers?

**Comment by BigLep** [comment]
@ajnavarro : I moved over the estimates from https://github.com/ipfs/kubo/issues/9188#issuecomment-1220886223 to here.

Can you help with the issue cleanup here?  Specifically:
1. I think we need a new issue in go-libp2p-routing-helpers for the new routers.
2. I assume we'll update https://github.com/ipfs/kubo/issues/9079 to account for the design changes.  This is "Make this work in Kubo" above right or are you going to create another issue?
3. I assume https://github.com/ipfs/kubo/issues/9079 and https://github.com/ipfs/kubo/issues/9083 can be closed in favor of the work happening in #2, is that right?

**Comment by ajnavarro** [comment]
- New issue for `parallel` and `sequential` routers https://github.com/libp2p/go-libp2p-routing-helpers/issues/56
- Updated issue https://github.com/ipfs/kubo/issues/9079
- What issue are you referring to? I think it is a typo.

**Comment by BigLep** [comment]
@ajnavarro : thanks.  I updated "# 3" above.  I believe the confusion stemmed from the fact that I wrote "#2" which got translated to issue "#2" in the repo.

**Comment by ajnavarro** [comment]
@BigLep yep, we can close #9083

**Comment by BigLep** [comment]
As [communicated in Slack](https://filecoinproject.slack.com/archives/C03RQFURZLM/p1660955676555979):

Total estimate is 13 ideal dev days
Assuming a velocity of .7 ideal dev day to actual day, that puts us at 19 calendar days assuming just one person working on this.
That feels high to me, but I think Bedrock should plan around that for now given we'll likely experience a jolt when Adin heads out, there is ongoing Kubo maintenance efforts that the team gets pulled into, and I don't think we'll be able to have an extra engineer on this beyond helping with code reviews.

**Comment by ajnavarro** [comment]
@BigLep I think we can do them separately.

**Comment by ajnavarro** [comment]
- Config parser with tests: Done
- Create Router instances from validated config: Done
- Integrate with fx: In progress

**Comment by ajnavarro** [comment]
- First draft PR uploaded: https://github.com/ipfs/kubo/pull/9274. 
- Waiting for a new review on https://github.com/libp2p/go-libp2p-routing-helpers/pull/58.
- Now updating go-delegated-routing to the latest release. There were some changes in method signatures.

**Comment by lidel** [comment]
I was not sure where "productizing" `https://routing.delegate.ipfs.io/reframe` should live, so I filled https://github.com/protocol/bifrost-infra/issues/2142

**Comment by BigLep** [comment]
@guseggert : I added the important area to the done criteria we were missing last week of "ipfs.io gateway operators are able to assess reframe call health with metrics".  Feel free to create a separate tracking issue for this work or do it here.

**Comment by BigLep** [comment]
I've expanded the done checklist to account for the migration from reframe to the updated HTTP delegated routing API: https://github.com/ipfs/go-delegated-routing/issues/63 .  I have updated the title as well.

**Comment by BigLep** [comment]
2023-04-25 maintainer conversation: 
we're still trying to confirm that the ipfs.io gateway is calling cid.contact.  
cid.contact is not seeing any requests from ipfs.io gateways per discussion in https://github.com/ipni/storetheindex/pull/1597#issuecomment-1519923480 and https://filecoinproject.slack.com/archives/C03RQFURZLM/p1682348755904599 .  
@guseggert SSH'd to a box and confirmed it is using "autoclient"
@guseggert confirmed that when he uses the same version of Kubo with "autoclient" that cid.contact is being called
@guseggert is leading the investigation here

**Comment by willscott** [comment]
I suspect https://github.com/ipfs/boxo/blob/main/routing/http/client/transport.go#L24-L28 could be an issue:
* if the response body is not fully drained, the http client will not be able to re-use the connection. eventually you'll run out of the pool of clients and I think the logic then is to fail silently.
* The defaults will give a pretty limited number of default allowed "simultaneously inflight" to a single host - that likely needs to be tuned for gateways. https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/net/http/transport.go;l=201-211

**Comment by guseggert** [comment]
> if the response body is not fully drained, the http client will not be able to re-use the connection. eventually you'll run out of the pool of clients and I think the logic then is to fail silently.

IIUC the body is not fully drained in the case where the limit is reached, which is indeed a bug, but are there actually records in cid.contact with 1 MiB of providers?

**Comment by guseggert** [comment]
Seems like a design flaw.  @willscott do you have any suggestions for mitigations?  An obvious one is to sacrifice perf for availability by waiting for a resp from cid.contact before returning, even when using FullRT, which will increase latency for most requests.

**Comment by willscott** [comment]
* there may be records with >1MiB of response

> bitswap issues a query for 10 providers and the accelerated DHT returns 20 immediately, terminating the query before it even starts on cid.contact.

I don't think this explains things - we have cases of providers that only are publishing to cid.contact.
there wouldn't be 20 peers already in the DHT there, and we don't find them on gateways.

Just because you have some peers from FullRT you don't need to close the channel / cancel the query on the indexer side, right?

(in the same way that when indexers gets requests requesting a cascade to the DHT they don't cancel the dht cascade once we've returned responses from the index.)

**Comment by guseggert** [comment]
Okay there's a plumbing bug in Kubo's rats nest of content routing plumbing that is causing this issue, when the experimental DHT client is turned on then the cid.contact router is not used. Working on a fix, could take me a few hours to untangle that mess.

**Comment by BigLep** [comment]
Thanks @guseggert for the update.  

As you are working on this, friendly reminder that we plan to move "accelerated DHT" out of experimental in the next release.  If there are followups that need to occur, fee free to drop them to https://github.com/ipfs/kubo/issues/9703

In terms of releasing this, I assume we'll have it part of 0.20?  (We'll talk during 2023-04-27 standup. Alternative is that Bifrost uses a custom branch until this come in a future release.)

A challenge here will be code review given team members being out.  I think our best case is that @lidel take a look at it Thursday, 2023-04-27.

**Comment by willscott** [comment]
i would push to get an RC onto bifrost and validate things are working as expected before releasing

**Comment by guseggert** [comment]
Agreed we should one-box this on the gateways before releasing to make sure it works, otherwise the feedback loop is brutal

**Comment by BigLep** [comment]
2023-04-27 maintainer conversation: this is still the top priority of @guseggert that he is actively working on.  We don't have a PR yet.

**Comment by guseggert** [comment]
Okay I have working code now and can confirm that the default HTTP routers are being invoked alongside the FullRT client, will open a PR shortly and the plan is for @Jorropo to take a look on Monday

**Comment by BigLep** [comment]
Thanks @guseggert .  Could also give directions on what Bifrost would need to test it?  I'd like to get their test/verification happening in parallel.

**Comment by BigLep** [comment]
2023-05-02 update: The fix has deployed to "one box" on ipfs.io gateway.  Integration looks good.  These are going to get rolled out to the whole fleet over the next day or two.

**Comment by BigLep** [comment]
I'm closing this issue because ipfs.io gateway <> cid.contact integration is rolled out (see https://github.com/protocol/bifrost-infra/issues/2183#issuecomment-1533342956) and confirmed to be working.
